### PR TITLE
feat: fix package exports and simplify build

### LINKED ISSUE #694: Broken exports
@thebuilder , something changed in between 9.10.3 and 9.12.0 that broke the existing imports:

Getting the following error for all my tests:
```
TS2307: Cannot find module 'react-intersection-observer/test-utils' or its corresponding type declarations.
```

Screenshot 2024-07-11 at 10 40 00

### PR DESCRIPTION
**Pull request by RobinTail**
Fixes #694 , @thebuilder 

Screenshot 2024-07-11 at 11 23 26

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -104,6 +111,7 @@
     "react-dom": "^15.0.0 || ^16.0.0 || ^17.0.0 || ^18.0.0 || ^19.0.0"
   },
   "devDependencies": {
+    "@arethetypeswrong/cli": "^0.15.3",
```
Comment (RobinTail) [inline package.json:114]: this thing immediately checks the build

File: package.json [hunk 2]
```diff
@@ -30,10 +40,7 @@
   "packageManager": "pnpm@9.5.0+sha256.dbdf5961c32909fb030595a9daa1dae720162e658609a8f92f2fa99835510ca5",
   "scripts": {
     "prebuild": "rm -rf dist lib",
-    "build": "run-s build:*",
-    "build:bundle": "tsup src/index.tsx",
-    "build:legacy": "tsup src/index.tsx --format esm --legacy-output --no-clean --no-dts",
-    "build:utils": "tsup src/test-utils.ts --no-clean --no-sourcemap",
```
Comment (RobinTail) [inline package.json:36]: all those commands are moved into a properly configured `tsup.config.ts`

File: package.json [hunk 3]
```diff
@@ -5,21 +5,31 @@
   "type": "commonjs",
   "source": "src/index.tsx",
   "main": "dist/index.js",
-  "module": "dist/esm/index.js",
+  "module": "dist/index.mjs",
   "types": "dist/index.d.ts",
   "exports": {
```
Comment (RobinTail) [inline package.json:10]: this entry does not work for "moduleResolution: node", that's why it's important to make "test-utils" a directory with "index.js" file inside it. All that is configured inside `tsup` config

File: tsup.config.ts [hunk 4]
```diff
@@ -1,11 +1,24 @@
...
+};
+
+export default defineConfig([
+  {
+    entryPoints: ["src/index.tsx"],
+    outDir: "dist",
+    ...commons,
+  },
+  {
+    entryPoints: { index: "src/test-utils.ts" },
```
Comment (RobinTail) [inline tsup.config.ts:20]: this syntax renames the file into "index" inside the assigned `outDir`

File: package.json (hunk 3 above)
Comment (thebuilder) [reply package.json:10]: This "hack" was done to fix another issue, with Webpack not understanding `.mjs` https://github.com/thebuilder/react-intersection-observer/issues/633

Comment (thebuilder) [reply package.json:10]: Modern tooling would use the `exports` field over the `module`, which is why it's sorta stuck in a limbo legacy hacks for ESM.

Comment (RobinTail) [reply package.json:10]: Ok. I can make a copy: "dist/esm/index.js" for that case, I think

Comment (RobinTail) [reply package.json:10]: `module` value restored in [a66dcae](https://github.com/thebuilder/react-intersection-observer/pull/695/commits/a66dcae3d347fd0be2d9f8129588e77e33f6a3fc), @thebuilder

**Comment by RobinTail** [comment]
What's failing in Vercel, @thebuilder ? Can't see it

**Comment by thebuilder** [comment]
> What's failing in Vercel, @thebuilder ? Can't see it

I made a PR on your fork - It's the the `storybook:build` task that tries to run the `build:bundle` task that was removed.

**Comment by RobinTail** [comment]
Needs your authorization to run Vercel check again

**Comment by RobinTail** [comment]
It seems to be working!
Wheeee! :)

**Comment by thebuilder** [comment]
Looks good! 🎉

**Comment by RobinTail** [comment]
Thank you for all the support and wonderful collaboration, @thebuilder ! ❤️

**Comment by thebuilder** [comment]
> Thank you for all the support and wonderful collaboration, @thebuilder ! ❤️

And thank you for going in and fixing up the build flow. It's been bugging me for a long time. Much cleaner now!

**Pull request merged**